janus.1
.TH JANUS 1
.SH NAME
janus \- WebRTC server/gateway
.SH SYNOPSIS
.B janus
[options]
.SH DESCRIPTION
.B janus
is a WebRTC server/gateway developed by Meetecho conceived to be a general purpose one. As such, it doesn't provide any functionality per se other than implementing the means to set up a WebRTC media communication with a browser or application, exchanging JSON messages with it over different transports, and relaying RTP/RTCP and messages between clients and the server-side application logic they're attached to. Any specific feature/application is provided by server side plugins, that browsers can then contact via the server to take advantage of the functionality they provide. Example of such plugins can be implementations of applications like echo tests, conference bridges, media recorders, SIP gateways and the like.
.TP
The reason for this is simple: we wanted something that would have a small footprint (hence a C implementation) and that we could only equip with what was really needed (hence pluggable modules). That is, something that would allow us to deploy either a full-fledged WebRTC server on the cloud, or a small nettop/box to handle a specific use case.
.SH OPTIONS
.TP
.BR \-h ", " \-\-help
Print help and exit
.TP
.BR \-V ", " \-\-version
Print version and exit
.TP
.BR \-b ", " \-\-daemon
Launch Janus in background as a daemon (default=off)
.TP
.BR \-p ", " \-\-pid-file=\fIpath\fR
Open the specified PID file when starting Janus (default=none)
.TP
.BR \-N ", " \-\-disable-stdout
Disable stdout based logging (default=off)
.TP
.BR \-L ", " \-\-log-stdout
Log to stdout, even when the process is daemonized (default=off)
.TP
.BR \-L ", " \-\-log-file=\fIpath\fR
Log to the specified file (default=stdout only)
.TP
.BR \-H ", " \-\-cwd-path=\fIpath\fR
Working directory for Janus daemon process (default=/)
.TP
.BR \-i ", " \-\-interface=\fIipaddress\fR
Interface to use (will be the public IP)
.TP
.BR \-P ", " \-\-plugins-folder=\fIpath\fR
Plugins folder (default=./plugins)
.TP
.BR \-C ", " \-\-config=\fIfilename\fR
Configuration file to use
.TP
.BR \-F ", " \-\-configs-folder=\fIpath\fR
Configuration files folder (default=./conf)
.TP
.BR \-c ", " \-\-cert-pem=\fIfilename\fR
DTLS certificate
.TP
.BR \-k ", " \-\-cert-key=\fIfilename\fR
DTLS certificate key
.TP
.BR \-K ", " \-\-cert-pwd=\fItext\fR
DTLS certificate key passphrase (if needed)
.TP
.BR \-S ", " \-\-stun-server=\fIaddress:port\fR
STUN server(:port) to use, if needed (e.g., Janus behind NAT, default=none)
.TP
.BR \-1 ", " \-\-nat-1-1=\fIips\fR
Comma-separated list of public IPs to put in all host candidates, assuming a 1:1 NAT is in place (e.g., Amazon EC2 instances, default=none)
.TP
.BR \-2 ", " \-\-keep-private-host
When nat-1-1 is used (e.g., Amazon EC2 instances), don't remove the private host, but keep both to simulate STUN (default=off)
.TP
.BR \-E ", " \-\-ice-enforce-list=\fIlist\fR
Comma-separated list of the only interfaces to use for ICE gathering; partial strings are supported (e.g., eth0 or eno1,wlan0, default=none)
.TP
.BR \-X ", " \-\-ice-ignore-list=\fIlist\fR
Comma-separated list of interfaces or IP addresses to ignore for ICE gathering; partial strings are supported (e.g., vmnet8,192.168.0.1,10.0.0.1 or vmnet,192.168., default=vmnet)
.TP
.BR \-6 ", " \-\-ipv6-candidates
Whether to enable IPv6 candidates or not (default=off)
.TP
.BR \-O ", " \-\-ipv6-link-local
Whether IPv6 link-local candidates should be gathered as well (default=off)
.TP
.BR \-l ", " \-\-libnice-debug
Whether to enable libnice debugging or not (default=off)
.TP
.BR \-f ", " \-\-full-trickle
Do full-trickle instead of half-trickle (default=off)
.TP
.BR \-I ", " \-\-ice-lite
Whether to enable the ICE Lite mode or not (default=off)
.TP
.BR \-T ", " \-\-ice-tcp
Whether to enable ICE-TCP or not (warning: only works with ICE Lite) (default=off)
.TP
.BR \-Q ", " \-\-min-nack-queue=\fInumber\fR
Minimum size of the NACK queue (in ms) per user for retransmissions, no matter the RTT
.TP
.BR \-t ", " \-\-no-media-timer=\fInumber\fR
Time (in s) that should pass with no media (audio or video) being received before Janus notifies you about this
.TP
.BR \-W ", " \-\-slowlink-threshold=\fInumber\fR
Number of lost packets (per s) that should trigger a 'slowlink' Janus API event to users (default=0, feature disabled)
.TP
.BR \-r ", " \-\-rtp-port-range=\fImin\-max\fR
Port range to use for RTP/RTCP
.TP
.BR \-B ", " \-\-twcc-period=\fInumber\fR
How often (in ms) to send TWCC feedback back to senders, if negotiated (default=200ms)
.TP
.BR \-n ", " \-\-server-name=\fIname\fR
Public name of this Janus instance (default=MyJanusInstance)
.TP
.BR \-s ", " \-\-session-timeout=\fInumber\fR
Session timeout value, in seconds (default=60)
.TP
.BR \-m ", " \-\-reclaim-session-timeout=\fInumber\fR
Reclaim session timeout value, in seconds (default=0)
.TP
.BR \-d ", " \-\-debug-level=\fI1\-7\fR
Debug/logging level (0=disable debugging, 7=maximum debug level; default=4)
.TP
.BR \-D ", " \-\-debug-timestamps
Enable debug/logging timestamps (default=off)
.TP
.BR \-o ", " \-\-disable-colors
Disable color in the logging (default=off)
.TP
.BR \-M ", " \-\-debug-locks
Enable debugging of locks/mutexes (very verbose! default=off)
.TP
.BR \-a ", " \-\-apisecret=\fIrandomstring\fR
API secret all requests need to pass in order to be accepted by Janus (useful when wrapping Janus API requests in a server, none by default)
.TP
.BR \-A ", " \-\-token-auth
Enable token-based authentication for all requests (default=off)
.TP
.BR \-e ", " \-\-event-handlers
Enable event handlers (default=off)
.TP
.BR \-w ", " \-\-no-webrtc-encryption
Disable WebRTC encryption, so no DTLS or SRTP (only for debugging!) (default=off)
.SH EXAMPLES
\fBjanus\fR \- Launch Janus with all options from configurations files
.TP
\fBjanus \-b \-L /tmp/januslog\fR \- Launch Janus as a daemon and log to the specified file
.TP
\fBjanus \-6\fR \- Launch Janus with IPv6 support enabled
.TP
\fBjanus \-f\fR \- Launch Janus with full-trickle enabled
.SH BUGS
.TP
If you think you found a bug or want to contribute a feature, you can issue or a pull request on https://github.com/meetecho/janus-gateway/issues.
.TP
Anyway, before doing that make sure you read the documentation at https://janus.conf.meetecho.com/docs/ and that it has not been discussed already at https://groups.google.com/forum/#!forum/meetecho-janus. We only use Github for code issues, and \fBNOT\fR for configuration or usage issues: use the group for that.
.SH SEE ALSO
.TP
https://github.com/meetecho/janus-gateway \- Official repository
.TP
https://janus.conf.meetecho.com \- Demos and documentation
.TP
https://groups.google.com/forum/#!forum/meetecho-janus \- Community
.TP
https://www.meetecho.com/blog/ \- Tutorials and blog posts on Janus
.SH AUTHORS
Lorenzo Miniero (lorenzo@meetecho.com)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.


## [v0.11.5] - 2021-10-18

- Add API to optionally force Janus to use TURN [[PR-2774](https://github.com/meetecho/janus-gateway/pull/2774)]
- Fixed slow path on SDP parsing [[PR-2776](https://github.com/meetecho/janus-gateway/pull/2776)]
- Added event handlers option to send stats for a PeerConnection in a single event, rather than per-media (thanks @JanFellner!) [[PR-2785](https://github.com/meetecho/janus-gateway/pull/2785)]
- Fixed occasional deadlocks on malformed requests in VideoRoom [[Issue-2780](https://github.com/meetecho/janus-gateway/issues/2780)]
- Fixed AudioBridge plain RTP thread sometimes exiting prematurely
- Fixed broken upsampling when using G.711 in AudioBridge
- Add pause/resume recording functionality to Record&Play and SIP plugins (thanks @isnumanagic!) [[PR-2724](https://github.com/meetecho/janus-gateway/pull/2724)]
- Fixed broken support for Unix Sockets in WebSockets Admin API (thanks @thatsmydoing!) [[PR-2787](https://github.com/meetecho/janus-gateway/pull/2787)]
- Added timing info for video rotation when post-processing recordings
- Added linter checks to janus.js (thanks @davel!) [[PR-2272](https://github.com/meetecho/janus-gateway/pull/2272)]
- Other smaller fixes and improvements (thanks to all who contributed pull requests and reported issues!)

## [v0.11.4] - 2021-09-06

- Fixed ICE restart issues with recent versions of libnice [[PR-2729](https://github.com/meetecho/janus-gateway/pull/2729)]
- Changed randon number generators to use crypto-safe functions (thanks @jmfotokite!) [[PR-2738](https://github.com/meetecho/janus-gateway/pull/2738)]
- Added support for abs-send-time RTP extension [[PR-2721](https://github.com/meetecho/janus-gateway/pull/2721)]
- Added configurable mechanism for manually setting static event loop to use for new handles [[PR-2684](https://github.com/meetecho/janus-gateway/pull/2684)]
- Fixed datachannel protocol not being sent to plugins for incoming messages [[Issue-2753](https://github.com/meetecho/janus-gateway/issues/2753)]
- Added ability to specify recordings folder in AudioBridge [[PR-2707](https://github.com/meetecho/janus-gateway/pull/2707)]
- Added support for forwarding groups in AudioBridge [[PR-2653](https://github.com/meetecho/janus-gateway/pull/2653)]
- Fixed missing Contact header in SIP plugin when using Sofia >= 1.13 [[PR-2708](https://github.com/meetecho/janus-gateway/pull/2708)]
- Better SDES-SRTP negotiation in SIP and NoSIP plugins [[PR-2727](https://github.com/meetecho/janus-gateway/pull/2727)]
- Fixed WebSocket transport and event handler lagging 25/30s when shutting down or reconnecting (thanks @JanFellner!) [[Issue-2734](https://github.com/meetecho/janus-gateway/issues/2734)]
- Fixed incoming_header_prefixes not working for helper sessions in SIP plugin
- Fix partial/broken ACL support in TextRoom plugin [[PR-2763](https://github.com/meetecho/janus-gateway/pull/2763)]
- Fixed potential race condition when reclaiming sessions in HTTP transport plugin
- Fixed WebSocket event handler reconnect mechanism (thanks @JanFellner!) [[PR-2736](https://github.com/meetecho/janus-gateway/pull/2736)]
- Other smaller fixes and improvements (thanks to all who contributed pull requests and reported issues!)

## [v0.11.3] - 2021-06-15

- Fixed rare crash when detaching handles [[Issue-2464](https://github.com/meetecho/janus-gateway/issues/2464)]
- Added option to offer IPv6 link-local candidates as well [[PR-2689](https://github.com/meetecho/janus-gateway/pull/2689)]
- Added spatial audio support to AudioBridge via stereo mixing [[PR-2446](https://github.com/meetecho/janus-gateway/pull/2446)]
- Added support for plain RTP participants to AudioBridge [[PR-2464](https://github.com/meetecho/janus-gateway/pull/2464)]
- Added API to start/stop AudioBridge recordings dynamically (thanks @rajneeshksoni!) [[PR-2674](https://github.com/meetecho/janus-gateway/pull/2674)]
- Fixed broken mountpoint switching when using different payload types in Streaming plugin [[PR-2692](https://github.com/meetecho/janus-gateway/pull/2692)]
- Fixed occasional deadlock on Streaming plugin mountpoint destroy during RTSP reconnects (thanks @lionelnicolas!) [[PR-2700](https://github.com/meetecho/janus-gateway/pull/2700)]
- Added "Expires" support to SUBSCRIBE in SIP plugin (thanks @nicolasduteil!) [[PR-2661](https://github.com/meetecho/janus-gateway/pull/2661)]
- Added option to specify Call-ID for SUBSCRIBE dialogs in SIP plugin (thanks @nicolasduteil!) [[PR-2664](https://github.com/meetecho/janus-gateway/pull/2664)]
- Fixed broken simulcast support in VideoCall plugin (thanks @lucily-star!) [[PR-2671](https://github.com/meetecho/janus-gateway/pull/2671)]
- Implemented RabbitMQ reconnection logic, in both transport and event handler (thanks @chriswiggins!) [[PR-2651](https://github.com/meetecho/janus-gateway/pull/2651)]
- Added support for renegotiation of external streams in janus.js (thanks @kmeyerhofer!) [[PR-2604](https://github.com/meetecho/janus-gateway/pull/2604)]
- Added support for HEVC/H.265 aggregation packets (AP) to janus-pp-rec (thanks @nu774!) [[PR-2662](https://github.com/meetecho/janus-gateway/pull/2662)]
- Refactored janus-pp-rec to cleanup the code, and use libavformat for Opus as well (thanks @lu-zero!) [[PR-2665](https://github.com/meetecho/janus-gateway/pull/2665)]
- Added additional target formats for some recorded codecs [[PR-2680](https://github.com/meetecho/janus-gateway/pull/2680)]
- Other smaller fixes and improvements (thanks to all who contributed pull requests and reported issues!)

## [v0.11.2] - 2021-05-03

- Added support for relative paths in config files, currently only in MQTT event handler (thanks @RSATom!) [[PR-2623](https://github.com/meetecho/janus-gateway/pull/2623)]

npm/janus.d.ts
declare namespace JanusJS {
	interface Dependencies {
		adapter: any;
		WebSocket: (server: string, protocol: string) => WebSocket;
		isArray: (array: any) => array is Array<any>;
		extension: () => boolean;
		httpAPICall: (url: string, options: any) => void;
	}

	interface DependenciesResult {
		adapter: any;
		newWebSocket: (server: string, protocol: string) => WebSocket;
		isArray: (array: any) => array is Array<any>;
		extension: () => boolean;
		httpAPICall: (url: string, options: any) => void;
	}

	enum DebugLevel {
		Trace = 'trace',
		Debug = 'debug',
		Log = 'log',
		Warning = 'warn',
		Error = 'error'
	}

	interface JSEP {}

	interface InitOptions {
		debug?: boolean | 'all' | DebugLevel[];
		callback?: Function;
		dependencies?: DependenciesResult;
	}

	interface ConstructorOptions {
		server: string | string[];
		iceServers?: RTCIceServer[];
		ipv6?: boolean;
		withCredentials?: boolean;
		max_poll_events?: number;
		destroyOnUnload?: boolean;
		token?: string;
		apisecret?: string;
		success?: Function;
		error?: (error: any) => void;
		destroyed?: Function;
	}

	interface ReconnectOptions {
		success?: Function;
		error?: (error: any) => void;
	}

	enum MessageType {
		Recording = 'recording',
		Starting = 'starting',
		Started = 'started',
		Stopped = 'stopped',
		SlowLink = 'slow_link',
		Preparing = 'preparing',
		Refreshing = 'refreshing'
	}

	interface Message {
		result?: {
			status: MessageType;
			id?: string;
			uplink?: number;
		};
		error?: Error;
	}

	interface PluginOptions {
		plugin: string;
		opaqueId?: string;
		dataChannelOptions?: RTCDataChannelInit;
		success?: (handle: PluginHandle) => void;
		error?: (error: any) => void;
		consentDialog?: (on: boolean) => void;
		webrtcState?: (isConnected: boolean) => void;
		iceState?: (state: 'connected' | 'failed') => void;
		mediaState?: (medium: 'audio' | 'video', receiving: boolean, mid?: number) => void;
		slowLink?: (state: { uplink: boolean }) => void;
		onmessage?: (message: Message, jsep?: JSEP) => void;
		onlocalstream?: (stream: MediaStream) => void;
		onremotestream?: (stream: MediaStream) => void;
		ondataopen?: Function;
		ondata?: Function;
		oncleanup?: Function;
		detached?: Function;
	}

	interface OfferParams {
		media?: {
			audioSend?: boolean;
			audioRecv?: boolean;
			videoSend?: boolean;
			videoRecv?: boolean;
			audio?: boolean | { deviceId: string };
			video?:
				| boolean
				| { deviceId: string }
				| 'lowres'
				| 'lowres-16:9'
				| 'stdres'
				| 'stdres-16:9'
				| 'hires'
				| 'hires-16:9';
			data?: boolean;
			failIfNoAudio?: boolean;
			failIfNoVideo?: boolean;
			screenshareFrameRate?: number;
		};
		trickle?: boolean;
		stream?: MediaStream;
		success: Function;
		error: (error: any) => void;
	}

	interface PluginMessage {
		message: {
			request: string;
			[otherProps: string]: any;
		};
		jsep?: JSEP;
		success?: Function;
		error?: (error: any) => void;
	}

	interface PluginHandle {
		plugin: string;
		id: string;
		token?: string;
		detached : boolean;
		webrtcStuff: {
			started: boolean,
			myStream: MediaStream,
			streamExternal: boolean,
			remoteStream: MediaStream,
			mySdp: any,
			mediaConstraints: any,
			pc: RTCPeerConnection,
			dataChannelOptions: RTCDataChannelInit,
			dataChannel: Array<RTCDataChannel>,
			dtmfSender: any,
			trickle: boolean,
			iceDone: boolean,
			volume: {
				value: number,
				timer: number
			}
		};
		getId(): string;
		getPlugin(): string;
		send(message: PluginMessage): void;
		createOffer(params: any): void;
		createAnswer(params: any): void;
		handleRemoteJsep(params: { jsep: JSEP }): void;
		dtmf(params: any): void;
		data(params: any): void;
		isAudioMuted(): boolean;
		muteAudio(): void;
		unmuteAudio(): void;
		isVideoMuted(): boolean;
		muteVideo(): void;
		unmuteVideo(): void;
		getBitrate(): string;
		hangup(sendRequest?: boolean): void;
		detach(params: any): void;
	}

	class Janus {
		static webRTCAdapter: any;
		static safariVp8: boolean;
		static useDefaultDependencies(deps: Partial<Dependencies>): DependenciesResult;
		static useOldDependencies(deps: Partial<Dependencies>): DependenciesResult;
		static init(options: InitOptions): void;
		static isWebrtcSupported(): boolean;
		static debug(...args: any[]): void;
		static log(...args: any[]): void;
		static warn(...args: any[]): void;
		static error(...args: any[]): void;
		static randomString(length: number): string;
		static attachMediaStream(element: HTMLMediaElement, stream: MediaStream): void;
		static reattachMediaStream(to: HTMLMediaElement, from: HTMLMediaElement): void;

		constructor(options: ConstructorOptions);

		getServer(): string;
		isConnected(): boolean;
		getSessionId(): string;
		attach(options: PluginOptions): void;
		reconnect(options: ReconnectOptions): void;
		destroy(): void;
	}
}

export default JanusJS.Janus;
export { JanusJS };
